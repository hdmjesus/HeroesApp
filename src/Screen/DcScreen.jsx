import React from "react";
import { HeroesList } from "../Components/heroes/List";

export const DcScreen = () => {
  return (
    <div className="container mt-5">
      <h1 className="m-3">Dc Heroes</h1>
      <hr />
      <HeroesList publisher={"DC Comics"} />
    </div>
  );
};
