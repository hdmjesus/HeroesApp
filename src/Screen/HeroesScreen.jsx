import React from "react";
import { Redirect, useParams } from "react-router";
import { getHeroById } from "../selectors/getHeroesById";

export const HeroesScreen = ({ history }) => {
  const { heroid } = useParams();
  const hero = getHeroById(heroid);
  console.log(hero);
  if (!hero) {
    return <Redirect to="/" />;
  }
  const { id, superhero, publisher, alter_ego, first_appearance, characters } =
    hero;
  const handleReturn = () => {
    if (history.length <= 2) {
      publisher === "Marvel Comics" && history.push("/");
      publisher === "DC Comics" && history.push("/dc");
    } else {
      history.goBack();
    }
  };
  return (
    <div className="animate__animated animate__fadeInLeft p-7 md:flex">
      <div className="">
        <img className=" " src={`../images/${heroid}.webp`} alt={heroid} />
      </div>
      <div className="container  flex mt-4 flex-col items-center md:items-start   ">
        <h3 className="md:ml-28 text-lg uppercase font-bold">{superhero}</h3>
        <ul className="list-group list-group-flush text-xs md:text-base md:m-auto md:mb-5 md:mt-5">
          <li className="list-group-item">
            {" "}
            <b>Alter Ego: </b> {alter_ego}
          </li>
          <li className="list-group-item">
            {" "}
            <b>Publisher: </b> {publisher}
          </li>
          <li className="list-group-item">
            {" "}
            <b>First Apearance: </b> {first_appearance}
          </li>
          <li className="list-group-item">
            {" "}
            <b>Characters: </b> {characters}
          </li>
        </ul>

        <button
          className="btn btn-outline-info mt-3 m-auto"
          onClick={handleReturn}
        >
          Return
        </button>
      </div>
    </div>
  );
};
