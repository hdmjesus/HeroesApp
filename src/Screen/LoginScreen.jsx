import React from "react";

export const LoginScreen = ({ history }) => {
  const handleLogin = () => {
    history.replace("/");
  };
  return (
    <div className="container mt-5">
      <h1 className="m-3">LoginScreen</h1>
      <hr />
      <button className="btn btn-primary m-4" onClick={handleLogin}>
        Login
      </button>
    </div>
  );
};
