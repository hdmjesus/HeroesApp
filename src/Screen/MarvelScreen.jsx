import React from "react";
import { HeroesList } from "../Components/heroes/List";

export const MarvelScreen = () => {
  return (
    <div className="container mt-5">
      <h1 className="m-3">Marvel Comics</h1>
      <hr />

      <HeroesList publisher={"Marvel Comics"} />
    </div>
  );
};
