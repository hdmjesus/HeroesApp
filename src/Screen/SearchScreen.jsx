import React, { useMemo, useState } from "react";
import queryString from "query-string";
import { useLocation } from "react-router";
import { HeroeCard } from "../Components/heroes/HeroeCard";
import { useForm } from "../Hooks/useForm";
import { getHeroesByName } from "../selectors/getHeroesByName";
export const SearchScreen = ({ history }) => {
  const [error, setError] = useState(false);
  const location = useLocation();
  const { q = " " } = queryString.parse(location.search);

  const [formValue, handleInputChange, reset] = useForm({ searchText: q });
  const { searchText } = formValue;

  const heroesFiltered = useMemo(() => getHeroesByName(searchText), [q]);

  const handleError = () => {
    if (q === "" || q === " ") {
      setError(false);
      console.log("error false");
    }
    if (heroesFiltered.length === 0 && q !== "") {
      setError(true);
      console.log(" error true");
    }
  };
  const handleSearch = (e) => {
    e.preventDefault();
    handleError();
    history.push(`?q=${searchText}`);
  };

  return (
    <div>
      <div className=" flex gap-5 flex-col md:flex-row container  mt-4">
        <div className="md:w-5/12">
          <h4 className="m-3">Search your heroe</h4>
          <hr />

          <form
            className="container mt-3 flex flex-col"
            action=""
            onSubmit={handleSearch}
          >
            <input
              className="form-control outline-none "
              type="text"
              name="searchText"
              id=""
              placeholder="Find you hero"
              onChange={handleInputChange}
              autoComplete="off"
              value={searchText}
            />
            <button
              type="submit"
              className="btn w-2/4 m-auto  md:w-full mt-3 btn-block btn-outline-primary"
            >
              Search
            </button>
          </form>
        </div>

        <div className="md:w-8/12 mt-4 ">
          {" "}
          <h4 className="mv-3">Results</h4>
          <hr />
          {q === " " && error == false ? (
            <div className="alert alert-info w-full m-2"> Search a heroe</div>
          ) : null}
          {heroesFiltered.length === 0 && !error && (
            <div className="alert alert-danger m-2">
              <p>
                Error, personaje <b>{q}</b> no econtrado
              </p>
            </div>
          )}
          <div className="flex flex-wrap m-auto">
            {heroesFiltered.map((hero) => (
              <HeroeCard key={hero.id} {...hero} />
            ))}
          </div>
        </div>
      </div>
    </div>
  );
};
