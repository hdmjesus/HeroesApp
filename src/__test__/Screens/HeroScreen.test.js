import React from "react";
import { mount } from "enzyme";
import { HeroesScreen } from "../../Screen/HeroesScreen";
import { MemoryRouter } from "react-router";
import { Route } from "react-router-dom";

describe("Pruebas en <HeroScreen/>", () => {
  const historyMock = { length: 10, push: jest.fn(), goBack: jest.fn() };

  test("debe mostrar el componente redirect si no hay algumentos en el url", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero"]}>
        <HeroesScreen history={historyMock} />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("Redirect").exists()).toBe(true);
  });

  test("debe de mostrar un heroe si el parametro existe y se encuentra  ", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero/dc-batman"]}>
        <Route path="/hero/:heroid" component={HeroesScreen} />
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".animate__animated").exists()).toBe(true);
  });

  test("debe hacer push si el history.length es menor a 2", () => {
    const historyMock = { length: 1, push: jest.fn(), goBack: jest.fn() };

    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero/dc-batman"]}>
        <Route
          path="/hero/:heroid"
          component={() => <HeroesScreen history={historyMock} />}
        />
      </MemoryRouter>
    );
    wrapper.find("button").simulate("click");
    expect(historyMock.push).toHaveBeenCalledWith("/dc");
    expect(historyMock.goBack).not.toHaveBeenCalled();
  });

  test("debe regresar a la pantalla anterior con goBack", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/hero/dc-batman"]}>
        <Route
          path="/hero/:heroid"
          component={() => <HeroesScreen history={historyMock} />}
        />
      </MemoryRouter>
    );

    wrapper.find("button").simulate("click");
    expect(historyMock.goBack).toHaveBeenCalled();
    expect(historyMock.push).not.toHaveBeenCalledWith("/dc");
  });
});
