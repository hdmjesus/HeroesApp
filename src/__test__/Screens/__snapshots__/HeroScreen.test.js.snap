// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HeroScreen/> debe de mostrar un heroe si el parametro existe y se encuentra   1`] = `
<div
  className="animate__animated animate__fadeInLeft p-7 md:flex"
>
  <div
    className=""
  >
    <img
      alt="dc-batman"
      className=" "
      src="../images/dc-batman.webp"
    />
  </div>
  <div
    className="container  flex mt-4 flex-col items-center md:items-start   "
  >
    <h3
      className="md:ml-28 text-lg uppercase font-bold"
    >
      Batman
    </h3>
    <ul
      className="list-group list-group-flush text-xs md:text-base md:m-auto md:mb-5 md:mt-5"
    >
      <li
        className="list-group-item"
      >
         
        <b>
          Alter Ego: 
        </b>
         
        Bruce Wayne
      </li>
      <li
        className="list-group-item"
      >
         
        <b>
          Publisher: 
        </b>
         
        DC Comics
      </li>
      <li
        className="list-group-item"
      >
         
        <b>
          First Apearance: 
        </b>
         
        Detective Comics #27
      </li>
      <li
        className="list-group-item"
      >
         
        <b>
          Characters: 
        </b>
         
        Bruce Wayne
      </li>
    </ul>
    <button
      className="btn btn-outline-info mt-3 m-auto"
      onClick={[Function]}
    >
      Return
    </button>
  </div>
</div>
`;

exports[`Pruebas en <HeroScreen/> debe mostrar el componente redirect si no hay algumentos en el url 1`] = `""`;
